zcash_client_backend/src/data_api/testing/pool/dsl.rs

```diff
@@ -0,0 +1,379 @@
+            .with_account_from_sapling_activation(BlockHash([0; 32]))
+            .into()
+    }
+
+    /// Build the builder, wrapping the resulting [`TestState`] in a [`TestDsl`] and [`TestScenario`].
+    pub fn build<T: ShieldedPoolTester>(self) -> TestDsl<TestScenario<T, Cache, Dsf>> {
+        let state = self.inner.build();
+        TestScenario {
+            state,
+            _phantom: PhantomData,
+        }
+        .into()
+    }
+}
+
+/// A proxy for `FakeCompactOutput` that allows test code to omit the `fvk` and
+/// `address_type` fields, which can be derived from the `TestState` in most cases.
+pub struct TestNoteConfig<T: ShieldedPoolTester> {
+    /// The amount of the note.
+    pub value: Zatoshis,
+    /// Diversifiable full viewing key of the recipient.
+    pub fvk: Option<T::Fvk>,
+    /// Address type of the recipient.
+    pub address_type: Option<AddressType>,
+}
+
+impl<T: ShieldedPoolTester> From<Zatoshis> for TestNoteConfig<T> {
+    fn from(value: Zatoshis) -> Self {
+        TestNoteConfig {
+            value,
+            fvk: None,
+            address_type: None,
+        }
+    }
+}
+
+impl<T: ShieldedPoolTester> TestNoteConfig<T> {
+    pub fn with_address_type(mut self, address_type: AddressType) -> Self {
+        self.address_type = Some(address_type);
+        self
+    }
+
+    pub fn with_fvk(mut self, fvk: T::Fvk) -> Self {
+        self.fvk = Some(fvk);
+        self
+    }
+}
+
+pub struct AddFundsStepResult<T: ShieldedPoolTester, C: TestCache> {
+    pub block_height: BlockHeight,
+    pub insert_result: C::InsertResult,
+    /// Empty when the step was to generate an empty block
+    pub nullifiers: Vec<<T::Fvk as TestFvk>::Nullifier>,
+}
+
+/// The input and output of one "add funds" step.
+pub struct AddFundsStep<T: ShieldedPoolTester, C: TestCache> {
+    pub notes: Vec<TestNoteConfig<T>>,
+    pub results: AddFundsStepResult<T, C>,
+}
+
+/// A collection of results from adding funds to a `TestState`.
+pub struct AddFundsSummary<T: ShieldedPoolTester, C: TestCache> {
+    pub steps: Vec<AddFundsStep<T, C>>,
+    pub scan_summary: Option<ScanSummary>,
+}
+
+impl<T: ShieldedPoolTester, C: TestCache> Default for AddFundsSummary<T, C> {
+    fn default() -> Self {
+        Self {
+            steps: Default::default(),
+            scan_summary: None,
+        }
+    }
+}
+
+impl<T: ShieldedPoolTester, C: TestCache> AddFundsSummary<T, C> {
+    /// Return the first block height.
+    pub fn first_block_height(&self) -> BlockHeight {
+        if let Some(step) = self.steps.first() {
+            step.results.block_height
+        } else {
+            BlockHeight::from_u32(0)
+        }
+    }
+
+    /// Return the latest block height after generating the blocks
+    /// that added funds.
+    pub fn block_height(&self) -> BlockHeight {
+        if let Some(step) = self.steps.last() {
+            step.results.block_height
+        } else {
+            BlockHeight::from_u32(0)
```

Comment:
I would have both of these return `None`. Also, do we not have `BlockHeight::GENESIS` or `BlockHeight::ZERO`? We should.